apps/omg/lib/omg/output.ex

```diff
@@ -19,56 +19,75 @@ defmodule OMG.Output do
 
       {:ok, _unrecognized_type} ->
         {:error, :unrecognized_output_type}
     end
   end
 
-  def dispatching_reconstruct(_), do: {:error, :malformed_outputs}
+  @doc """
+  Reconstructs the structure from a list of RLP items
+  """
+  def reconstruct([output_type, [owner_rlp, currency_rlp, amount_rlp]]) do
+    with {:ok, cur12} <- RawData.parse_address(currency_rlp),
+         {:ok, owner} <- RawData.parse_address(owner_rlp),
+         :ok <- non_zero_owner(owner),
+         {:ok, amount} <- RawData.parse_amount(amount_rlp),
+         do: %__MODULE__{output_type: output_type, owner: owner, currency: cur12, amount: amount}
+  end
 
-  def from_db_value(%{type: output_type} = db_value), do: @output_types_modules[output_type].from_db_value(db_value)
-  # default clause for backwards compatibility
-  def from_db_value(%{} = db_value), do: OMG.Output.FungibleMoreVPToken.from_db_value(db_value)
-end
+  def reconstruct(_), do: {:error, :malformed_outputs}
 
-defprotocol OMG.Output.Protocol do
-  @moduledoc """
-  Captures the varying behavior of outputs that build the plasma chain
+  defp non_zero_owner(<<0::160>>), do: {:error, :output_guard_cant_be_zero}
+  defp non_zero_owner(_), do: :ok
 
-  Includes the "output predicate", within the `can_spend?/3` function
-  """
+  def dispatching_reconstruct(_), do: {:error, :malformed_outputs}
 
-  @doc """
-  True if a particular witness can unlock a particular output to be spent, given being put in a particular transaction
+  def from_db_value(%{type: output_type} = db_value), do: @output_types_modules[output_type].from_db_value(db_value)
 
-  Intended to be called in stateful validation
-  """
-  def can_spend?(output_spent, witness, raw_tx)
+  def from_db_value(%{owner: owner, currency: currency, amount: amount, output_type: output_type})
+      when is_binary(owner) and is_binary(currency) and is_integer(amount) and is_integer(output_type) do
+    %__MODULE__{owner: owner, currency: currency, amount: amount, output_type: output_type}
+  end
 
   @doc """
-  Returns the input pointer that the output should be later referenced by in inputs to be spent
+  For payment outputs, a binary witness is assumed to be a signature equal to the payment's output owner
   """
-  def input_pointer(output, blknum, tx_index, oindex, tx, hash)
+  def can_spend?(%__MODULE__{owner: owner}, witness, _raw_tx) when is_binary(witness) do
+    owner == witness
+  end
 
-  @doc """
-  Transforms into a db-specific term
-  """
-  def to_db_value(output)
+  def input_pointer(%__MODULE__{}, blknum, tx_index, oindex, _, _),
+    do: Utxo.position(blknum, tx_index, oindex)
 
-  @doc """
-  Transforms into a RLP-ready structure
-  """
-  def get_data_for_rlp(output)
+  def to_db_value(%__MODULE__{owner: owner, currency: currency, amount: amount, output_type: output_type})
+      when is_binary(owner) and is_binary(currency) and is_integer(amount) and is_integer(output_type) do
```

Comment:
I don't understand these guards? who are they for? what are they protecting us from?